audio/snd-util/adagio/mp.1
.TH MP 1 LOCAL
.SH NAME
mp \- midi player
.SH SYNOPSIS
.B mp
[ options ] file [ options ]
.SH DESCRIPTION
.I Mp
plays a
midi
file on sound cards under
.IR Linux ,
and/or sends
midi
commands out through the sound card's
external port.
Alternatively, the
midi
file can be converted to an adagio score
(with the
.B \-a
flag).
.PP
If a file with the name supplied on the command line is not
found,
.I mp
looks for a file by that name with the suffix ".mid".
Option flags may precede or follow the file name, and single
letter option flags may be combined.
.PP
.I Mp
requires the sound driver by Hannu Savolainen for access
to sound cards or a midi port.
For sound cards (SoundBlaster, PAS16, GUS), instrument
patches must be loaded into the sound driver before a
midi file is played.
These patches and information about them are in files which
must be in known locations, so that
.I mp
can find them.
The directories where
.I mp
will look for these files are configurable at compile time,
so the specific directories mentioned below are merely the
defaults -- they may have been changed.
.PP
Libraries of patches for sound cards are kept in /etc/std.* and
and /etc/drums.*, where the "*" is a suffix telling the
type of patch: "sb" for 2 operator fm cards, "o3" for
4 operator fm cards, and "gus" for the Gravis Ultrasound.
Actually, the files /etc/std.gus and /etc/drums.gus contain
only some information about GUS patches, which are too bulky
to be kept in libraries.  The GUS patches themselves are
kept in either the directory /dos/ultrasnd/midi/ or the
directory /dos/ultrasnd/midi2/ (the latter is searched first).
.PP
In addition to patches and libraries of patches, certain
optional configuration information can be kept in cfg files.
A file /etc/default.cfg, if it exists, is read before playing
every file.  This can be used to alter the default GUS or fm
patch assignments, to prevent playing on one of the
available devices, and for other purposes -- there is a
separate document, CFG-FILES, describing cfg files.
Other cfg files can contain information about playing
specific files or about playing groups of files in a
specific directory.
.PP
For the GUS, mp can use special patches for tone banks
and alternate drum sets, but the information about these
patches has to be supplied in a cfg file (see CFG-FILES).
.PP
Both format 0 and format 1 multi-track files can be played.
The midi damper pedal controller and, for sound cards, the
chorus depth and reverberation controllers are simulated in
software.
Midi main volume and expression controls are approximated by modifying
patch parameters and volume for the sound cards and the Kawai
K1, and passed along to the external port if it is in use for
other external synths.
Other controllers are also passed along to the external port, but
ignored for playing on the cards.
.PP
The choice of which channels will be passed to the external port is
a compilation option, unless you happen to have a Kawai K1 synth
attached, in which case
.I mp
can make sensible choices about which channels to send to it, and can
also approximate the effect of the pan controller by allocating
voices to left, center, or right positions.
.PP
There are the following command line options:
.TP
.BR \-adagio \ (or \ \-a )
send an adagio score to stdout, rather than playing.
.TP
.BR \-f
means don't play anything on the fm card.
.TP
.BR \-g
means don't play anything on the gus.
.TP
.BR \-external \ (or \ \-e )
means don't send any channels to the external midi port.
(Which channels will be external, if any, in the absence of
this flag, depends on a compilation option -- see the Makefile.)
.TP
.BR \-p
plays all parts except percussion with piano, ignoring program
requests.
.TP
.BR \-r \ chan[,prog]
records midi input from the external port in a new file "mpout.mid".
"Chan" is a number from 1-16, the channel for the midi commands in
the file, and "prog" is a number from 0-127, for the general midi
program to be used.  The output midi file contains a single track
which can be merged with the file being played by the separate
utility programs
.IR midt \ and \ tm .
.TP
.BR \-R \ reverb
adds reverberation to all channels; "reverb" is a number from 1-127.
.TP
.BR \-C \ depth
adds chorus depth to all channels; "depth" is a number from 1-127.
.TP
.BR \-s
prevents treating any instruments as "solo" or non-polyphonic instruments.
A new note on a solo instrument cuts off any previous note (e.g.,
on a flute as opposed to piano).
.TP
.BR \-init \ (or \ \-i )
prevents a series of initialization commands for
each channel from being issued at the beginning of
the track.
.TP
.B \-help
displays a usage summary.
.TP
.B \-v
gives some information about text embedded in the midi file and
which voices are used for which midi channels.  Voice numbers
and tone banks are shown using the notation [voice-number,tone-bank].
.TP
.B \-vv
gives even more information than
.B \-v
does.
.TP
.BR \-tune \ file
uses tuning information from the file.
.TP
.BR \-trace \ (or \ \-t )
issues tracing information.
.SH "SEE ALSO"
ad(1), setfm(1), midt(1), tracks(1), ``The CMU Midi Toolkit'' by Roger Dannenberg
.SH BUGS
Slow to start up.  There is no general mechanism for assigning channels
and programs to output devices.
.SH AUTHORS
Roger Dannenberg,
Roger.Dannenberg@SPICE.CS.CMU.EDU,
(Adagio and Transcribe interpreter code),
.br
Greg Lee, lee@uhunix.uhcc.hawaii.edu
(the rest).


audio/snd-util/adagio/mp.c
/*
 * mp (Midi Play)
 *
 * This program is largely derived from a program
 * 'mftext' by Tim Thompson, and programs 'transcribe'
 * and 'adagio' by Roger Dannenberg.  Some of the code
 * for playing the SoundBlaster card came from 'fmplay'
 * by Hannu Savolainen (hsavolai@cs.helsinki.fi).
 *
 * First, a list of midi events is built up by call-backs
 * by midifile to "txt_" functions.  Second, this list of
 * midi events is used to contruct a list of notes and
 * controls, which is, third, passed on to phase2 for
 * playing.
 *
 * For the midifile functions, see midifile.c and midifile.3.
 * For the playing functions, see phase2.c.
 *
 * (Building two intermediate representations makes the
 * program slow, but it let me use Dannenberg's code
 * with least modification.)
 *
 * (The neat function titles and comments bordered by lots
 * of asterisks are from Dannenberg.)
 *		Greg Lee, lee@uhunix.uhcc.hawaii.edu
 *		1/30/93
 */


#include <stdio.h>
#include <ctype.h>
#define NO_LC_DEFINES
#include "midifile.h"


#include "adagio.h"
#include "userio.h"
#include "cmdline.h"
#include "allphase.h"

void phase2();

/****************************************************************************
* Variables set by command line switches
****************************************************************************/

int ad_print = false;		/* adagio output */
int vverbose = false;		/* tracing output */

/****************************************************************************
* Data for command line parsing
****************************************************************************/
#define nswitches 18
static char *switches[nswitches] =
{"-help", "-v", "-vv",
 "-init", "-i",
 "-trace", "-t",
 "-external", "-e",
 "-midi", "-m",
 "-adagio", "-a",
 "-p", "-s",
 "-f", "-g", "-d"};

#define noptions 4
static char *options[noptions] =
{"-tune", "-r", "-R", "-C"};
#define n_t_sw 2
static char *t_switches[n_t_sw] =
{"-t", "-trace"};
#define n_a_sw 2
static char *a_switches[n_a_sw] =
{"-a", "-adagio"};
#define n_e_sw 2
static char *e_switches[n_e_sw] =
{"-e", "-external"};


/****************************************************************************
*	Routines local to this module
****************************************************************************/
static void cmdline_help();

/****************************************************************************
*				 cmdline_help
* Effect:
*	Prints out command line help
****************************************************************************/

static void cmdline_help()
{
    fprintf(stderr, "mp [options] filename [options]\n");
    fprintf(stderr, "	   -help	     this message\n");
    fprintf(stderr, "	   -init (-i)        initialize channels\n");
    fprintf(stderr, "	   -v    	     verbose\n");
    fprintf(stderr, "	   -vv   	     very verbose\n");
    fprintf(stderr, "	   -tune file	     use tuning from file\n");
    fprintf(stderr, "	   -trace (-t)       trace music\n");
    fprintf(stderr, "	   -adagio (-a)      adagio file on stdout\n");
    fprintf(stderr, "	   -p                piano only\n");
    fprintf(stderr, "	   -s                no solo instruments\n");
    fprintf(stderr, "	   -r  chan[,prog]   record new track\n");
    fprintf(stderr, "	   -e                no external synth\n");
    fprintf(stderr, "	   -g                no gus synth\n");
    fprintf(stderr, "	   -f                no fm synth\n");
    fprintf(stderr, "	   -d                no drum rolls\n");
    fprintf(stderr, "	   -R  reverb        midi reverberation\n");
    fprintf(stderr, "	   -C  chdepth       midi chorus depth\n");
}


/****************************************************************************
*	Routines in phasem.c
****************************************************************************/
void var_init();
void initfuncs();
void rec_init();
void card_init(char *devname);
event_type rec_final();
/*************************/

/* part of interface to midifile functions */
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

static unsigned char *midi_file_contents;
static int midi_file_size;

static int filegetc()
{
    if (midi_file_size-- > 0) return (*midi_file_contents++);
    return(EOF);
}

void txt_error(char *s)
{
    fprintf(stderr, "midifile error: %s\n", s);
    exit(1);
}

long *shm_running_time;
static long keep_running_time;

int main(argc, argv)
int argc;
char **argv;
{
    char *midi_part_name;
    char midi_name[100];
    struct stat statbuf;
    int fd;

    if (!cl_init(switches, nswitches, options, noptions, argv, argc))
	exit(1);
    if (cl_switch("-help")) {
	cmdline_help();
	return (0);
    }

    if ((midi_part_name = cl_arg(1)) == NULL) {
	fprintf(stderr, "please give a file name\n");
	exit(1);
    }
    strcpy(midi_name, midi_part_name);
    if (stat(midi_name, &statbuf)) {
	strcat(midi_name, ".mid");
	if (stat(midi_name, &statbuf)) {
	    perror(midi_name);
	    exit(1);
	}
    }
    strcpy(midi_file_path, midi_name);
    midi_file_size = statbuf.st_size;
    if ( (fd = open(midi_name, O_RDONLY, 0)) == -1 ) {
	perror(midi_name);
	exit(1);
    }
    if ( (midi_file_contents = (unsigned char *)malloc(midi_file_size)) == NULL) {
	perror("malloc");
	exit(1);
    }
    if (read(fd, midi_file_contents, midi_file_size) != midi_file_size) {
	perror("read");
	exit(1);
    }
    close(fd);
#ifdef READ_MODS
    Mf_file_contents = midi_file_contents;
    Mf_file_size = midi_file_size;
#endif

    shm_running_time = &keep_running_time;

    var_init();


    if (cl_switch("-v")) verbose = 1;
    if (cl_switch("-vv")) {
	really_verbose = 1;
	verbose = 1;
    }
    if (cl_switch("-d")) setting_drum_rolls = 0;
    piano_only = cl_switch("-p");
    exclude_fm = cl_switch("-f");
    exclude_gus = cl_switch("-g");
    no_solo = cl_switch("-s");
    extflag = (cl_nswitch(e_switches, n_e_sw) == NULL);
    ad_print = (cl_nswitch(a_switches, n_a_sw) != NULL);
    vverbose = (cl_nswitch(t_switches, n_t_sw) != NULL);
    if (cl_option("-r") != NULL)
	recording_track = true;


    card_init("/dev/sequencer");/* open /dev/sequencer and get info on devices */
    rec_init();
    initfuncs();		/* set calls to us from midifile functions */
    Mf_getc = filegetc;		/* tell midifile how to get bytes to process */
    mfread();			/* call midifile */
    free(midi_file_contents);
    if (ad_print)		/* Just translating to Adagio score? */
	(void) rec_final(true);	/* write out recorded data, */
    /* suppress time of first event*/
    else
	phase2((rec_final(true))); /* No, we're actually going to play it. */
    exit(0);
}
